xclim/core/base.py

```diff
@@ -0,0 +1,407 @@
+        """
+        return get_local_attrs(
+            self, locale, fill_missing=fill_missing, append_locale_name=False
+        )
+
+    @property
+    def cf_attrs(self):
+        """CF-Convention attributes of the output value."""
+        attrs = {k: getattr(self, k) for k in self._cf_names if getattr(self, k)}
+        return attrs
+
+    def json(self, args=None):
+        """Return a dictionary representation of the class.
+
+        Notes
+        -----
+        This is meant to be used by a third-party library wanting to wrap this class into another interface.
+
+        """
+        names = ["identifier", "var_name", "abstract", "keywords"]
+        out = {key: getattr(self, key) for key in names}
+        out.update(self.cf_attrs)
+        out = self.format(out, args)
+
+        out["notes"] = self.notes
+
+        out["parameters"] = str(
+            {
+                key: {
+                    "default": p.default if p.default != p.empty else None,
+                    "desc": "",
+                }
+                for (key, p) in self._sig.parameters.items()
+            }
+        )
+
+        # if six.PY2:
+        #     out = walk_map(out, lambda x: x.decode('utf8') if isinstance(x, six.string_types) else x)
+
+        return out
+
+    def format(
+        self,
+        attrs: dict,
+        args: dict = None,
+        formatter: AttrFormatter = default_formatter,
+    ):
+        """Format attributes including {} tags with arguments.
+
+        Parameters
+        ----------
+        attrs: dict
+          Attributes containing tags to replace with arguments' values.
+        args : dict
+          Function call arguments.
+        """
+        if args is None:
+            return attrs
+
+        out = {}
+        for key, val in attrs.items():
+            mba = {"indexer": "annual"}
+            # Add formatting {} around values to be able to replace them with _attrs_mapping using format.
+            for k, v in args.items():
+                if isinstance(v, dict):
+                    if v:
+                        dk, dv = v.copy().popitem()
+                        if dk == "month":
+                            dv = "m{}".format(dv)
+                        mba[k] = dv
+                elif isinstance(v, units.Quantity):
+                    mba[k] = "{:g~P}".format(v)
+                elif isinstance(v, (int, float)):
+                    mba[k] = "{:g}".format(v)
+                else:
+                    mba[k] = v
+
+            if callable(val):
+                val = val(**mba)
+
+            out[key] = formatter.format(val, **mba)
+
+            if key in self._text_fields:
+                out[key] = out[key].strip().capitalize()
+
+        return out
+
+    @staticmethod
+    def cfprobe(self, *das):
+        """Check input data compliance to expectations.
+        Warn of potential issues."""
+        return True
+
+    @staticmethod
+    def compute(self, *args, **kwds):
```

Comment:
I'm having a hard time understanding how this works: you are declaring a `staticmethod` but still having a `self` argument? Isn't the point of static methods to be independent of the object instantiation?

Reply:
Well, that's a bug.